evm/src/arithmetic/byte.rs

```diff
@@ -0,0 +1,502 @@
+//!
+//! -*-
+//!
+//! To prove that the bytes x and y are in the range [0, 2^8) (rather
+//! than [0, 2^16), which is all the range-checker guarantees) we do
+//! the following (hat tip to Jacqui for this trick too): Instead of
+//! storing x and y, we store w = 256 * x and y. Then, to verify that
+//! x, y < 256 and the last limb L = x + y * 256, we check that
+//! L = w / 256 + y * 256.
+//!
+//! The proof of why verifying that L = w / 256 + y * 256
+//! suffices is as follows:
+//!
+//! 1. The given L, w and y are range-checked to be less than 2^16.
+//! 2. y * 256 ∈ {0, 256, 512, ..., 2^24 - 512, 2^24 - 256}
+//! 3. w / 256 = L - y * 256 ∈ {-2^24 + 256, -2^24 + 257, ..., 2^16 - 2, 2^16 - 1}
+//! 4. By inspection, for w < 2^16, if w / 256 < 2^16 or
+//!    w / 256 >= P - 2^24 + 256 (i.e. if w / 256 falls in the range
+//!    of point 3 above), then w = 256 * m for some 0 <= m < 256.
+//! 5. Hence w / 256 ∈ {0, 1, ..., 255}
+//! 6. Hence y * 256 = L - w / 256 ∈ {-255, -254, ..., 2^16 - 1}
+//! 7. Taking the intersection of ranges in 2. and 6. we see that
+//!    y * 256 ∈ {0, 256, 512, ..., 2^16 - 256}
+//! 8. Hence y ∈ {0, 1, ..., 255}
+
+use ethereum_types::U256;
+use plonky2::field::extension::Extendable;
+use plonky2::field::packed::PackedField;
+use plonky2::field::types::{Field, PrimeField64};
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use static_assertions::const_assert;
+
+use crate::arithmetic::columns::*;
+use crate::arithmetic::utils::u256_to_array;
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+
+// Give meaningful names to the columns of AUX_INPUT_REGISTER_0 that
+// we're using
+const BYTE_LAST_LIMB_LO: usize = AUX_INPUT_REGISTER_0.start + 7;
+const BYTE_LAST_LIMB_HI: usize = AUX_INPUT_REGISTER_0.start + 8;
+const BYTE_IDX_IS_LARGE: usize = AUX_INPUT_REGISTER_0.start + 9;
+const BYTE_IDX_HI_LIMB_SUM_INV_0: usize = AUX_INPUT_REGISTER_0.start + 10;
+const BYTE_IDX_HI_LIMB_SUM_INV_1: usize = AUX_INPUT_REGISTER_0.start + 11;
+const BYTE_IDX_HI_LIMB_SUM_INV_2: usize = AUX_INPUT_REGISTER_0.start + 12;
+const BYTE_IDX_HI_LIMB_SUM_INV_3: usize = AUX_INPUT_REGISTER_0.start + 13;
+
+/// Decompose `idx` into bits and bobs and store in `idx_decomp`.
+///
+/// Specifically, write
+///
+///   idx = idx0_lo5 + idx0_hi * 2^5 + \sum_i idx[i] * 2^(16i),
+///
+/// where `0 <= idx0_lo5 < 32` and `0 <= idx0_hi < 2^11`.  Store the
+/// 5 bits of `idx0_lo5` in `idx_decomp[0..5]`; we don't explicitly need
+/// the higher 11 bits of the first limb, so we put them in
+/// `idx_decomp[5]`. The rest of `idx_decomp` is set to 0.
+fn set_idx_decomp<F: PrimeField64>(idx_decomp: &mut [F], idx: &U256) {
+    for i in 0..5 {
+        idx_decomp[i] = F::from_bool(idx.bit(i));
+    }
+    idx_decomp[5] = F::from_canonical_u16((idx.low_u64() as u16) >> 5);
+    for i in 6..N_LIMBS {
+        idx_decomp[i] = F::ZERO;
+    }
+}
```

Comment:
I would make this function accept a slice of length 6, not of length `N_LIMBS`. That way you don't have to worry about the remaining limbs.